### PR TITLE
Allow to use a dynamic attribute on the . operator via ()

### LINKED ISSUE #3598: attribute() not looking for camelCase method
Hi,
I think it's a sort of bug, but it may be classified as a whish.

I'm using Symfony. I've added the non public attribute $event_login to an entity class. 
I meet a twig error when using {{ attribute(item, field) }} : Neither the property "event_login" nor one of the methods "event_login()", "getevent_login()"/"isevent_login()"/"hasevent_login()" or "__call()" exist and have public access in class "App\Entity\Event".
So I add the getevent_login() getter ...
Then I meet Symfony\Component\PropertyAccess\Exception\NoSuchPropertyException : Can't get a way to read the property "event_login" in class "App\Entity\Event".
So I add the getEventLogin() getter ...

Why cannot they use the same getter ?

After a few exchanges, man states that a method should always been camelCase() (see https://www.php-fig.org/psr/psr-1/ ).
Note that using underscores in attributes is allowed.
That's why PropertyAccess use getEventLogin() ... 

So, why does the twig attribute() function use getevent_login() instead of getEventLogin() ?

Should not it at least also check for getEventLogin() existence ?

Regards

### PR DESCRIPTION
**Pull request by fabpot**
Closes #3598

As one can think that the `attribute` function has some special power compared to the `.` operator (like using Symfony PropertyAccess), I think it's better to use the `.` operator for dynamic attributes as well.

Using `()` makes it clear that the behavior is the same as the `.` operator (using `()` is similar to the way one can use dynamic mapping names (like in `{ (name): value }`).

### CODE REVIEW

File: CHANGELOG [hunk 1]
```diff
@@ -1,5 +1,6 @@
 # 3.15.0 (2024-XX-XX)
 
+ * Deprecate the `attribute` function; use the `.` notation and wrap the name with parenthesis instead
```
Comment (stof) [inline CHANGELOG:3]: Deprecating the `attribute` function would make it hard to write Twig templates supporting older Twig versions (we cannot do a `{% if %}` to switch between 2 cases, we have to switch between 2 templates instead as the deprecation is at compile time).

As the `attribute` function is free to maintain (the implementation uses the same PHP function than the `.` operator), I suggest delaying the deprecation until some future 4.x release (we have some maintained versions of Symfony that still support Twig 2.x and so could not upgrade).

Comment (fabpot) [reply CHANGELOG:3]: That works for me. Code updated accordingly.

Comment (fabpot) [reply CHANGELOG:3]: See 8d1af0fc03314a2786decd6457447a583f7dd146

File: doc/functions/attribute.rst [hunk 2]
```diff
@@ -1,6 +1,13 @@
 ``attribute``
 =============
 
+.. warning::
+
+    The ``attribute`` filter is deprecated as of Twig 3.15. Use the ``.``
```
Comment (stof) [inline doc/functions/attribute.rst]: should this be a link to the dot operator documentation ?

Comment (fabpot) [reply doc/functions/attribute.rst]: Good idea, done

**Comment by mishoo** [comment]
I just noticed this in the documentation, and I was wondering why the weird syntax? Instead of `{{ user.('get' ~ name) }}`, wouldn't this work: `user['get' ~ name]`?

Asking because I maintain a Twig→JS compiler, and I'm pondering whether to implement the new syntax.. it seems superfluous to me.

**Comment by stof** [comment]
@mishoo both work to access an array index. But `[ ]` does not allow calling a method (in JS, calling a method is about reading a property containing a function and calling it, but that's absolutely not the case in PHP)

**Pull request merged**